### ISSUE TITLE
PythonEval doesn't build with python3

### ISSUE THREAD

**Issue by ebadawy** [opencog/atomspace#501]
I've a problem with atomespace build, got this error with `make`

```
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc: In function ‘bool try_to_load_modules(const char**)’:
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc:227:73: error: ‘PyString_AsString’ was not declared in this scope
             const char* sysPathCString = PyString_AsString(pySysPathLine);
                                                                         ^
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc: In function ‘void opencog::global_python_initialize()’:
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc:302:42: error: cannot convert ‘char**’ to ‘wchar_t**’ for argument ‘2’ to ‘void PySys_SetArgv(int, wchar_t**)’
         PySys_SetArgv(1, (char **) &argv0);
                                          ^
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc: In member function ‘void opencog::PythonEval::build_python_error_message(const char*, std::string&)’:
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc:570:66: error: ‘PyString_AsString’ was not declared in this scope
         char* pythonErrorString = PyString_AsString(pyErrorString);
                                                                  ^
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc: In member function ‘void opencog::PythonEval::execute_string(const char*)’:
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc:607:18: error: ‘Py_FlushLine’ was not declared in this scope
     Py_FlushLine();
                  ^
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc: In member function ‘int opencog::PythonEval::argument_count(PyObject*)’:
/home/ebadawy/git/OpenCog/atomspace/opencog/cython/PythonEval.cc:621:57: error: ‘PyInt_AsLong’ was not declared in this scope
             argumentCount = PyInt_AsLong(pyArgumentCount);
                                                         ^
opencog/cython/CMakeFiles/PythonEval.dir/build.make:62: recipe for target 'opencog/cython/CMakeFiles/PythonEval.dir/PythonEval.cc.o' failed
make[2]: *** [opencog/cython/CMakeFiles/PythonEval.dir/PythonEval.cc.o] Error 1
CMakeFiles/Makefile2:1462: recipe for target 'opencog/cython/CMakeFiles/PythonEval.dir/all' failed
make[1]: *** [opencog/cython/CMakeFiles/PythonEval.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

I'm using arch linux with Python 3.5.0

**Comment by linas** [comment]
opencog requires python2. I'm surprised that cmake didn't check the python version.

Someone really needs to port it to python 3...

**Comment by ehab93** [comment]
Thanks @linas , it builded successfully with python2, but I think it will be a good move if you make it works with python3 as well.

**Comment by linas** [comment]
The original bug is 3 years old:  https://github.com/opencog/opencog/issues/48 -- someone needs to add python3 support

**Comment by ehab93** [comment]
Hey @linas , I've now a build error with opencog, got this:
```
[ 92%] Building CXX object opencog/visualization/gtk-visualizer/CMakeFiles/OpenCogAtomSpaceVisualizer.dir/AtomSpaceInterface.cpp.o
/home/ebadawy/git/OpenCog/opencog/opencog/visualization/gtk-visualizer/AtomSpaceInterface.cpp: In member function ‘std::string AtomSpaceInterface::PostToCogServer(std::string, std::string)’:
/home/ebadawy/git/OpenCog/opencog/opencog/visualization/gtk-visualizer/AtomSpaceInterface.cpp:356:6: error: no match for ‘operator<<’ (operand types are ‘std::ostream {aka std::basic_ostream<char>}’ and ‘std::istream {aka std::basic_istream<char>}’)
  cout<<response_stream<<endl;
      ^
```
And a long long list of notes.

Steps that I've done so far:
1. made a virtualenv to use python2
2. run install-archlinux-dependencies.sh
3. get to opencog dir and do the build steps there, then I got the above error with `make`.

Any idea how to fix that? Thanks in advance.

**Comment by linas** [comment]
try replacing cout by std::cout ?

**Comment by linas** [comment]
the gtk-visualizer is probably obsolete. I never tried building it.

**Comment by ehab93** [comment]
No, I didn't try to replace any thing. Should I just run `sudo make install` and ignore that error?

**Comment by inflector** [comment]
How big a priority is Python 3 support?

**Comment by linas** [comment]
I dunno. Low. people assume you are an antiquated dinosaur if you don't support it.

**Comment by rTreutlein** [comment]
I have just pushed my python3 branch. It contains my work trying to get AtomSpace to work with Python3. There is at least 1 issue remaining that i was unable to solve.

**Comment by inflector** [comment]
@rTreutlein What did you call your python3 branch? I'm going to look at Python 3 soon.

**Comment by AmeBel** [comment]
https://github.com/opencog/atomspace/tree/python3

**Comment by inflector** [comment]
@AmeBel Thanks, I thought he was referring to a branch on his fork of opencog. Didn't realize it was a branch on the main opencog repository.

**Comment by inflector** [comment]
@linas @pennachin Do you think it will be fine to only support the latest Python 3 versions? For example, 3.3 and above or perhaps 3.4 and above? There are some provisions that might be easier in the later Pythons as some features were added to make code that attempts to run in both 2.6 and later of the 2.x line and 3.3 and later of the 3.3 line work well.

Or would it be acceptable to just run the version on Ubuntu 14.04 (which is 3.4.3) and newer?

**Comment by linas** [comment]
trusty and newer (3.4.3) seems quite acceptable.

**Comment by pennachin** [comment]
I agree, Trusty and newer for Python 2 and 3 sounds good...

**Comment by edpell** [comment]
Hi still no python3 support. at level 3.5.2

**Comment by chris-hawley** [comment]
Any word on python3 support?

**Comment by linas** [comment]
Someone's gotta do the work. We need someone with the time and energy to maintain the python bindings; besides python3, there are also a number of python interface bugs that need fixing.

**Comment by linas** [comment]
Fedora has been working hard to remove python2, an will achieve this  in 2020 
http://phoronix.com/scan.php?page=news_item&px=Fedora-Py3-Default-30s

**Comment by linas** [comment]
Closing, mostly fixed in #1519  If there are more issues, open a more bug reports.

**Comment by linas** [comment]
OK, so #1567 will change things to that python3 is built by default, if python3 is found, else fall back to python2.
